[DOCS] update HETERO execution (#10758)
openvinotoolkit/openvino: docs/OV_Runtime_UG/hetero_execution.md

+-------------------------------+--------------------------------------------+-----------------------------------------------------------+
| Parameter Name & C++ property | Property values                            | Description                                               |
+===============================+============================================+===========================================================+
| | "MULTI_DEVICE_PRIORITIES"   | | HETERO: <device names>                   | | Lists the devices available for selection.              |
| | `ov::device::priorities`    | | comma-separated, no spaces               | | The device sequence will be taken as priority           |
| |                             | |                                          | | from high to low.                                       |
+-------------------------------+--------------------------------------------+-----------------------------------------------------------+

> 

Reviewer: why have we moved the table at the beginning? Table is a dry reference information while affinities, auto / manual modes are more interesting for users to understand HETERO capabilities.

Reviewer: Moreover, the table is not rendered well. Please, download compiled version of the docs.

Author: The table hasn't moved but I wanted to combine information on auto / manual - some of it was before the table and the section with a header was below - I put the top in the below part :) I see no problem in switching places between the "policies for assigning affinities" and "configuring hetero."
And yes, obviously, I have removed the sphinx directive around the table. My mistake, corrected.

### Manual and Automatic policies for assigning affinities

> **The manual policy** assumes setting affinities explicitly for all operations in the model using `ov::Node::get_rt_info` with the `"affinity"` key. 

Reviewer: Manual mode is more complex, so the auto mode should come first in docs because it's applicable for most of the users.

Author: The manual mode description is shorter (even though the operation itself is more complex) and this solution works in every case. The automatic option's description is more complex, has some exceptions, and refers to the manual mode for cases in which it does not work. I think the reader will have a better picture reading this if they get the manual info before. And with so little text, the "manual" information might look as an afterthought when placed after the automatic one. Just a proposition though.

Reviewer: let's have a call

Author: sure thing, whenever you have time, just let me know.

       :language: python
       :fragment: [set_manual_affinities]

@endsphinxdirective

**The automatic fallback** policy decides which operation goes to which device automatically according to the support in dedicated devices (`GPU`, `CPU`, `MYRIAD`, etc.) and query model step is called implicitly by Hetero device during model compilation.

The automatic fallback policy causes "greedy" behavior and assigns all operations that can be executed on a given device to it, according to the priorities you specify (for example, `ov::device::priorities("GPU,CPU")`).
It does not take into account device peculiarities such as the inability to infer certain operations without other special operations placed before or after that layer. If the device plugin does not support the subgraph topology constructed by the HETERO device, then you should set affinity manually.

@sphinxdirective

.. tab:: C++

    .. doxygensnippet:: docs/snippets/ov_hetero.cpp
       :language: cpp
       :fragment: [compile_model]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_hetero.py
       :language: python
       :fragment: [compile_model]

@endsphinxdirective

> The fallback policy does not work if at least one operation has `"affinity"` initialized. If you want to adjust affinities which were set automatically, you need to fix them, which usually serves minimizing the number of total subgraphs to optimize memory transfers:

Author: ```suggestion
#### Using Manual and Automatic Modes in Combination
In some cases you may need to consider manually adjusting affinities which were set automatically. It usually serves minimizing the number of total subgraphs to optimize memory transfers. To do it, you need to "fix" the automatically assigned affinities like so:
```

| | "MULTI_DEVICE_PRIORITIES"   | | HETERO: <device names>                   | | Lists the devices available for selection.              |
| | `ov::device::priorities`    | | comma-separated, no spaces               | | The device sequence will be taken as priority           |
| |                             | |                                          | | from high to low.                                       |
+-------------------------------+--------------------------------------------+-----------------------------------------------------------+
@endsphinxdirective

### Manual and Automatic modes for assigning affinities

#### The Manual Mode
It assumes setting affinities explicitly for all operations in the model using `ov::Node::get_rt_info` with the `"affinity"` key. 

@sphinxdirective

.. tab:: C++

    .. doxygensnippet:: docs/snippets/ov_hetero.cpp
       :language: cpp
       :fragment: [set_manual_affinities]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_hetero.py
       :language: python
       :fragment: [set_manual_affinities]

@endsphinxdirective

> ### The Automatic Mode

Author: ```suggestion
#### The Automatic Mode
```

@endsphinxdirective

> > **NOTE**: `ov::Core::query_model` does not depend on affinities set by a user. Instead, it queries for an operation support based on device capabilities.

Author: ```suggestion
Importantly, the automatic mode will not work if any operation in a model has its `"affinity"` already initialized. 

> **NOTE**: `ov::Core::query_model` does not depend on affinities set by a user. Instead, it queries for an operation support based on device capabilities.
```

@endsphinxdirective

Importantly, the automatic mode will not work if any operation in a model has its `"affinity"` already initialized.

> Importantly, the automatic mode will not work if any operation in a model has its `"affinity"` already initialized. 

Reviewer: looks like the line in duplicated

> 
> Importantly, the automatic mode will not work if any operation in a model has its `"affinity"` already initialized. 

Author: ```suggestion
```

@endsphinxdirective

> In the example above, the `CPU` device is configured to enable profiling data. `GPU` has the configuration property to perform inference in `f16` precision, while CPU has the default execution precision.

Reviewer: looks like you mixed the devices. We need:
- profiling for GPU
- fp32 for CPU

Author: I have no idea how and why the discrepancy :) Thanks, corrected accordingly.

> In the example above, the `CPU` device is configured to enable profiling data. `GPU` has the configuration property to perform inference in `f16` precision, while CPU has the default execution precision.

Author: ```suggestion
In the example above, the `GPU` device is configured to enable profiling data and uses the default execution precision, while `CPU` has the configuration property to perform inference in `fp32`.
```
